src/base/ClpMap.h

```diff
@@ -0,0 +1,248 @@
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID__SRC_BASE_CLPMAP_H
+#define SQUID__SRC_BASE_CLPMAP_H
+
+#include "mem/PoolingAllocator.h"
+#include "sbuf/Algorithms.h"
+#include "SquidTime.h"
+
+#include <functional>
+#include <list>
+#include <unordered_map>
+
+template<class EntryValue>
+size_t
+DefaultMemoryUsage(const EntryValue *e)
+{
+    return sizeof(*e);
+}
+
+/**
+ * A Map for caching data by Capacity, Lifetime, and Priority (CLP)
+ * Unlike other Map containers data is;
+ * - added only if it fits within a predetermined memory limit (Capacity),
+ * - gets expired based on TTL (Lifetime), and a fading Priority Queue.
+ */
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *) = DefaultMemoryUsage>
+class ClpMap
+{
+public:
+    class Entry
+    {
+    public:
+        Entry(const Key &aKey, EntryValue *t, int ttl): key(aKey), value(t), expires(squid_curtime+ttl) {}
+        ~Entry() {delete value;}
+        Entry(Entry &&) = default;
+        Entry & operator = (Entry &&) = default;
+    private:
+        Entry(const Entry &) = delete;
+        Entry & operator = (const Entry &) = delete;
+    public:
+        Key key; ///< the key of entry
+        EntryValue *value = nullptr; ///< A pointer to the stored value
+        time_t expires = 0; ///< When the entry is to be removed
+    };
+
+    /// container for stored data
+    typedef std::list<Entry, PoolingAllocator<Entry> > Storage;
+    typedef typename Storage::iterator StorageIterator;
+
+    /// Key:Entry* mapping for fast lookups by key
+    typedef std::pair<Key, StorageIterator> MapItem;
+    /// key:queue_item mapping for fast lookups by key
+    typedef std::unordered_map<Key, StorageIterator, std::hash<Key>, std::equal_to<Key>, PoolingAllocator<MapItem> > KeyMapping;
+    typedef typename KeyMapping::iterator KeyMapIterator;
+
+    ClpMap(int ttl, size_t size);
+    ~ClpMap() = default;
+    /// Search for an entry, and return a pointer
+    EntryValue *get(const Key &key);
+    /// Add an entry to the map
+    bool add(const Key &key, EntryValue *t);
+    /// Add an entry to the map with specific TTL
+    bool add(const Key &key, EntryValue *t, int ttl);
+    /// Delete an entry from the map
+    bool del(const Key &key);
+    /// (Re-)set the maximum size for this map
+    void setMemLimit(size_t aSize);
+    /// The available size for the map
+    size_t memLimit() const {return memLimit_;}
+    /// The free space of the map
+    size_t freeMem() const { return (memLimit() > memoryUsed() ? memLimit() - memoryUsed() : 0);}
+    /// The current size of the map
+    size_t memoryUsed() const {return memUsed_;}
+    /// The number of stored entries
+    int entries() const {return entries_;}
+private:
+    ClpMap(ClpMap const &);
+    ClpMap & operator = (ClpMap const &);
+
+    bool expired(const Entry &e) const;
+    void trim(size_t wantSpace = 0);
```

Comment:
After other changes, the strange zero default would be gone:

```suggestion
    void trim(size_t wantSpace);
```

Reply:
Done in 9cd838d3c